crmsh/join_lock.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (C) 2020 Xin Liang <XLiang@suse.com>
+# See COPYING for license information.
+
+
+import os
+import re
+import time
+import signal
+import threading
+from contextlib import contextmanager
+
+from . import bootstrap
+from . import utils
+
+
+def signal_handler(signum, frame):
+    raise RuntimeError("Error in JoinLock update timestamp thread, exit!")
+
+
+class JoinLock(object):
+    """
+    Class to manage lock for multiple nodes join in parallel
+    """
+
+    JOIN_LOCK_DIR = "/usr/share/crmsh/.join_lock_directory"
+    JOIN_LOCK_FILE = "{}/join_lock_file".format(JOIN_LOCK_DIR)
+    TAIL_CMD = "tail -1 {}".format(JOIN_LOCK_FILE)
+    MKDIR_CMD = "mkdir {}".format(JOIN_LOCK_DIR)
+    RM_CMD = "rm -rf {}".format(JOIN_LOCK_DIR)
+    UPDATE_CMD = "echo {} `date +%s` >> {}".format(utils.this_node(), JOIN_LOCK_FILE)
+    DETECT_INTERVAL = 15
+    UPDATE_INTERVAL = 10
+    SSH_OPTION = "-o ConnectTimeout={} -o StrictHostKeyChecking=no".format(UPDATE_INTERVAL)
+
+    def __init__(self, target_node):
+        """
+        Init function
+        """
+        self.thread_stop_event = threading.Event()
+        self.target_node = target_node
+
+    def _run(self, cmd):
+        """
+        Run command on target node, consider specific exceptions
```

Comment:
in this function we should check for error code due to ssh.
The funcion should then imho return the error code of the `executed` command which is then  controlled by the place is executed.
So we replace this regex on output with err-code.

Example:

```python
  elif re.search("mkdir: cannot create directory .*: File exists", err):
                raise FileExistsError(err)
```
should be removed from this function and checked  in the appropriate place.

Now I personally dunno if `utils.get_stdout_stderr` function provides actually 2 return codes.
One of the `ssh` command and one for the `executed` command. this should imho the proper way to handle this